framework/src/Volo.Abp.UI.Navigation/Volo/Abp/Ui/Navigation/MenuManager.cs

```diff
@@ -32,21 +34,69 @@ public async Task<ApplicationMenu> GetAsync(string name)
                 {
                     await contributor.ConfigureMenuAsync(context);
                 }
+
+                await CheckPermissionsAsync(scope.ServiceProvider, menu);
             }
 
             NormalizeMenu(menu);
 
             return menu;
         }
 
+        protected virtual async Task CheckPermissionsAsync(IServiceProvider serviceProvider, IHasMenuItems menuWithItems)
+        {
+            var requiredPermissionItems = new List<ApplicationMenuItem>();
+            GetRequiredPermissionNameMenus(menuWithItems, requiredPermissionItems);
+
+            if (requiredPermissionItems.Any())
+            {
+                var permissionChecker = serviceProvider.GetRequiredService<IPermissionChecker>();
+                var grantResult = await permissionChecker.IsGrantedAsync(requiredPermissionItems.Select(x => x.RequiredPermissionName).ToArray());
+
+                var toBeDeleted = new List<ApplicationMenuItem>();
```

Comment:
`toBeDeleted` can be HashSet which has much faster `ContainsKey` compared to List.